### DISCUSSION TITLE
Euler Angles conversion to Alt-Az?

### DISCUSSION THREAD

**Discussion by ultrapre** [Stellarium/stellarium#3032] [General]
I meet the same problem like the guys [here](https://www.reddit.com/r/Astronomy/comments/3udenf/quaternion_matrix_or_euler_angles_conversion_to/). 
I want to use 9-axis gyroscope on alt-az telescope to become push-to device, and it gives roll, pitch, yaw  (x,y,z) angles as below. Assume all x,y,z =0 while  telescope is pointing to north, I want to calculate az-alt from  roll, pitch, yaw after rotated. That is point on how to convert formula.

![2023-01-30 15 21 51](https://user-images.githubusercontent.com/32613953/215412773-4f13c17a-8b6b-43ad-a5d2-8ea333878475.png)

Here comes a formula from alt-az to Eular, elevation is alt. But I donnot know whether is true, and how to write a reversed formula. 

(telescope is point to z-axis)
```
    yaw = arcsin(sin(elevation)*sin(azimuth))
    pitch = arctan2(sin(elevation)*cos(azimuth), cos(elevation))
    roll = arctan2(sin(azimuth), cos(azimuth)*cos(elevation))
```

**Comment by ultrapre** [comment]
@gzotti @alex-w

**Comment by 10110111** [comment]
I'd say roll=0, pitch=altitude, yaw=azimuth (up to a sign). This assumes that the RPY=0,0,0 corresponds to a look at North, with up direction of the image pointing to zenith. Another assumption is that roll, pitch, yaw rotations are applied in this exact order (gyro-stabilized gimbals often have other orders of axes).

**Reply by ultrapre** [reply]
When the telescope point at high alt, change of az will influence roll.

**Reply by 10110111** [reply]
Depends on the mount. If the first angle (the one closest to the telescope itself) that is controlled is roll, or if the roll is not controlled like in a Dobson mount, then it's fixed regardless of the other angles.

**Reply by ultrapre** [reply]
Unfortunately it is dobson. The key may be the x axis is always parallel to the horizontal plane. 
Push-to device is using hand pushing, I hope it can work well in random moving direction.

**Reply by 10110111** [reply]
Then it's a two-axis mount, and you have to abandon one of your inputs. The easiest is to abandon roll, in this case you'll still be able to point to any direction in the sky using, as I said above, pitch to control altitude and yaw to control azimuth. Without controlling roll you just get a picture rotated with respect to the desired roll angle.

**Reply by ultrapre** [reply]
So you meant I can only first move from (0,0,0) to the right az and next move up to right alt, each object need to  do so.

**Reply by 10110111** [reply]
Yes, this is how any alt-az mount works without a third axis.

**Comment by gzotti** [comment]
Yes, roll=0, pitch=altitude, yaw=azimuth. Did not check the formula in details. If "[x, y, z=0] means pointing to the North" means the north point on the horizon, i.e., azimuth=0, altitude=0, rot=0, Z is looking in view direction, and Y axis is pointing UP, then Z rotation angle should always remain zero, Y angle is azimuth, and X angle is altitude (or negatives.) Take a rectangular camera, attach it to the top of the scope, and check. It should look like AltAz mount (default view) in Stellarium.

**Reply by ultrapre** [reply]
It seems that the Z is always 0, but in fact, there is field rotation

**Reply by gzotti** [reply]
Yes, that's a common issue when photographing (or observing) through alt-az mounted telescopes.

**Comment by ultrapre** [comment]
Let's simulate it. The inclinometer is located at the center of the sphere, the x-axis pointing to the moving object on the spherical surface (spherical coordinates (r=1, φ, θ)), and the y-axis is always parallel to the horizontal plane, and the rotation angles of the x, y, z axes are roll, pitch, yaw respectively; when the spherical coordinates are (1, 0, 0), the rotation angle of the sensor is 0. According to roll, pitch, yaw, calculate φ, θ. It seems that for the only one (roll, pitch, yaw), there is a  certain φ, θ.

**Reply by 10110111** [reply]
Not quite. For any roll you get the same (φ, θ) pair. Spherical coordinates are agnostic of the rotation around the view axis.

**Reply by ultrapre** [reply]
It has been limited, 
> the y-axis is always parallel to the horizontal plane

More accurrately, antiaircraft gun? When you point to any point, roll, pitch and yaw will change.

**Reply by gzotti** [reply]
Please keep your axis orientations as defined previously. X is altitude axis ("parallel to the horizon"), Z (roll==0) is pointing towards target. Then Y axis=azimuth.

You rotate around Y in azimuth, then around X in altitude. Just like your neck and head. (Usually you keep it straight, and not tilt to the side to "rotate view around Z")

**Reply by gzotti** [reply]
Aiming at a single point (like a star) for viewing does not care for roll.

**Reply by ultrapre** [reply]
Yes, we does not care for roll, but when telescope rotates, `pitch=altitude, yaw=azimuth.` is not right. Assume we move like dobison to change az, it changes both roll and yaw.

**Reply by 10110111** [reply]
> It has been limited,

> > the y-axis is always parallel to the horizontal plane

This limitation is natural for spherical coordinates. How they work: first pitch is applied (which is equivalent to altitude), which raises or lowers the x axis. The y axis remains the same. Then yaw is applied (equivalent to ±azimuth) *to the whole platform*, which rotates the y axis in the horizontal plane, and also rotates the raised/lowered x axis around the z axis. That's all.

> More accurrately, antiaircraft gun?

Not sure how they are mounted, but I guess it's the same as the tank cannon. This is exactly what I've just described. If you attach a horizontal bar to the end of this cannon (perpendicularly to cannon's axis), it'll remain horizontal regardless of what you aim at.

**Reply by ultrapre** [reply]
We only have one 9-axis gyroscope to complete it. It is attached to telescope. It detect the Euler angles.

**Reply by 10110111** [reply]
> We only have one 9-axis gyroscope to complete it. It is attached to telescope. It detect the Euler angles.

Not a problem at all (though these are not proper Euler angles, but rather Tait-Bryan ones). Just make sure that its roll axis is aligned with the telescope axis, and the yaw axis is aligned with zenith when the telescope is horizontal. In this case it should always report roll near zero.

**Reply by ultrapre** [reply]
So the mpu-6050 and mpu-9250 etc will keep rolls=0 when telescope rotates without caring path and sequences?

**Reply by 10110111** [reply]
If attached correctly, then yes, they should, within error margin.
